pkg/lib/packagekit.es6

```diff
@@ -302,3 +309,172 @@ export function watchRedHatSubscription(callback) {
+ * install_missing_packages.  It contains these fields:
+ *
+ * - missing_names:     Packages that were requested, are currently not installed,
+ *                      and can be installed.
+ *
+ * - unavailable_names: Packages that were requested, are currently not installed,
+ *                      but can't be found in any repository.
+ *
+ * If unavailable_names is empty, a simulated installation of the missing packages
+ * is done and the result also contains these fields:
+ *
+ * - extra_names:       Packages that need to be installed as dependencies of
+ *                      missing_names.
+ *
+ * - remove_names:      Packages that need to be removed.
+ *
+ * - download_size:     Bytes that need to be downloaded.
+ *
+ * Both check_missing_packages and install_missing_packages have a
+ * "progress" callback parameter, like cancellableTransaction.
+ *
+ * In addition to the usual "waiting", "percentage", and "cancel" fields,
+ * the progress object reported by install_missing_packages also includes
+ * "info" and "package" from the "Package" signal.
+ */
+
+export function check_missing_packages(names, progress_cb) {
+    var data = {
+        missing_names: [ ],
+        unavailable_names: [ ],
+    };
+
+    if (names.length === 0)
+        return Promise.resolve(data);
+
+    function refresh() {
+        return cancellableTransaction("RefreshCache", [ false ], progress_cb);
+    }
+
+    function resolve() {
+        data.missing_ids = [ ];
+
+        var installed_names = { };
+
+        return cancellableTransaction("Resolve",
+                                      [ Enum.FILTER_ARCH |
+                                        Enum.FILTER_NOT_SOURCE |
+                                        Enum.FILTER_NEWEST,
+                                      names
+                                      ],
+                                      progress_cb,
+                                      {
+                                          Package: (info, package_id) => {
+                                              var parts = package_id.split(";");
+                                              var repos = parts[3].split(":");
+                                              if (repos.indexOf("installed") >= 0) {
+                                                  installed_names[parts[0]] = true;
+                                              } else {
+                                                  data.missing_ids.push(package_id);
+                                                  data.missing_names.push(parts[0]);
+                                              }
+                                          },
+                                      })
+                .then(() => {
+                    for (var i = 0; i < names.length; i++) {
+                        if (!installed_names[names[i]] && data.missing_names.indexOf(names[i]) == -1)
+                            data.unavailable_names.push(names[i]);
```

Comment:
Instead of accessing `names[i]` several times, how about `names.forEach(name => { ... })`?

Reply:
Done.